### PR TITLE
Add resource details panel

### LINKED ISSUE #1208: Resource details panel on dashboard
Our resources page is becoming information rich. At some point we should add a panel to show details of the resource rather than cluttering the main page by adding all data in the table. We would need to figure out what information we want user to see as overview, and what details are extra which can be shown on demand. This would also help with #567 where we can show extra information from custom resources which we don't know where to fit in existing columns in the panel for user to see.

cc: @leslierichardson95

### LINKED ISSUE #1489: Show endpoints even if they aren't http
We only show http endpoints in the dashboard now. We should show any endpoint but http endpoints should be hyperlinks.

### LINKED ISSUE #1584: Display short container ID
The full docker container ID is long. 64 characters. Ain't nobody got time for that.

Docker command line tools and UI tooling such as Docker for Windows display the short container ID. The short container ID is the first 12 characters of the ID, and realistically will always be enough to uniquely identify a container.

See https://www.baeldung.com/ops/docker-container-id-from-name

Examples:

```
$ docker container ls -a
CONTAINER ID   IMAGE          COMMAND                  CREATED          STATUS          PORTS     NAMES
80f1bc1e7feb   nginx:alpine   "/docker-entrypoint.…"   36 seconds ago   Up 36 seconds   80/tcp    web-server-11
acdea168264a   nginx:alpine   "/docker-entrypoint.…"   36 seconds ago   Up 36 seconds   80/tcp    web-server-10
0cbfc6c17009   nginx:alpine   "/docker-entrypoint.…"   37 seconds ago   Up 36 seconds   80/tcp    web-server-1
```

image

The Aspire dashboard should display the short ID. The copy button should copy the full 64-character value.

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspire/issues/1208
Fixes https://github.com/dotnet/aspire/issues/1489
Fixes https://github.com/dotnet/aspire/issues/1584

**READY FOR REVIEW**

Playing around adding a resource panel. The resource grid is trying to display too much information and a panel allows less used content to be accessed there and removed from the grid.

* Add details button to resources grid
* Remove environment variables button from resources grid
* Move logs button from resources grid to details panel
* Add high-level resource values to panel. These are from returned resource properties
  * There is the concept of "known properties". These are displayed by default and have a friendly name in UI (can still hover over text to get the actual key).
  * Unknown properties are hidden by default. These are properties that aren't generally useful to the end user, e.g. the resource UID
  * The hide/show button toggles showing unknown properties.
  * This pattern will work well for custom resource types. Custom properties can be viewed in the UI.
* Add resource endpoints/services to panel
* Refactor resources panel to use an accordion like span panel
* Remove various bits of detail from the resources grid
  * Container ID
  * Container ports
  * Container command/args
  * Process ID

Future work:
* Display the resource panel by clicking the row instead of a details button. Will do in a follow up PR.

Demo (**updated 2024/9/1**):
resources-details-new

~Note: The design is ready to review. Code is rough/unfinished in places. Will improve after design feedback.~

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1542)

### CODE REVIEW

**Comment by JamesNK** [comment]
cc @leslierichardson95 @davidfowl

File: src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor
Comment (drewnoakes) [inline src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor:81]:
```suggestion
                @if (FilteredItems.Any())
```

File: src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs
Comment (drewnoakes) [inline src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs:55]: If the name is for display, it should be localised.

File: src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs
Comment (drewnoakes) [inline src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs:59]: If the name is for display, it should be localised.

**Review by drewnoakes — COMMENTED** [review]
Please hold off on merging this. The classes you're using in the front end will no longer exist once https://github.com/dotnet/aspire/pull/1476 is merged. I have new code that will produce the name/values you're producing here too, in a way that's extensible for arbitrary resource types.

**Comment by leslierichardson95** [comment]
I like the resources panel!  +1 on Adam's comment about there being duplicate info though.  While there may be some use cases for having duplicate fields, I worry that that may result in taking up more real estate on screen than needed.

**Comment by JamesNK** [comment]
Now that the resource refactor has happened, I think the way this UI works could be changed a bit. Each resource has a raw list of properties. To properly handle custom resources with unknown properties, this UI should display a list of properties.

I think by default the UI:
* Displays known properties for known resource types (e.g. container ID for containers, source path for projects, etc) or all properties if the resource type is unknown
* Clicking show all then shows all properties

File: src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs
Comment (JamesNK) [reply src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs]: Will do localization once UI is agreed on.

File: src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs
Comment (JamesNK) [reply src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs]: Will do localization once UI is agreed on.

**Comment by drewnoakes** [comment]
If you want to give any of the properties custom treatment, we can model the required metadata for that in the `ResourceType` messages we send to the client.

**Comment by davidfowl** [comment]
What's this change waiting on?

**Comment by JamesNK** [comment]
I made large refactors yesterday. Me finishing that work. Then moving hardcoded UI text to resource files.

**Comment by JamesNK** [comment]
This is now ready for final review.

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/StateColumnDisplay.razor
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/ResourcesGridColumns/StateColumnDisplay.razor:28]: I made some changes to the unread errors component because I wanted to reuse it in the resource details component.

At least for now it isn't reused but I left the changes in because I think they're A Good Improvement.

**Comment by drewnoakes** [comment]
Just pulled this down to take it for a spin. It's the end of my day so I'll look more tomorrow.

Initial impression is that it's a bit tricky to parse the grouping between sections. For example:

image

Adding some indentation here might help:

image

Another idea is to add a background to the expander headers.

**Comment by davidfowl** [comment]
I like the background idea.

**Comment by kvenkatrajan** [comment]
The new details panel is looking great. Approved.

I have gone ahead and added another issue for tracking the visibility on the expanded headers in lieu of not further increasing the size and scope of this PR : #1612

@adamint  please do an accessibility pass after merge and add another issue if additional work needs to be done there. Thanks!

**Review by kvenkatrajan — APPROVED** [review]

**Comment by JamesNK** [comment]
The accordion originally had a lot more stuff to separate sections. Different background colors for headers, borders, etc. I turned it off because it didn't look good.

I've slightly increased the gap between each section to help highlight them:
image

**Comment by JamesNK** [comment]
From offline: Will merge after @drewnoakes takes a look at the code.

File: src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs
Comment (drewnoakes) [inline src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs:91]: We shouldn't need to hard-code known properties for each resource type here. The `ResourceProperty` message received via gRPC includes a display name field. The dashboard should avoid special knowledge about specific fields wherever possible.

It also means we don't have to keep this code up to date as properties are added/removed.

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/SourceColumnDisplay.razor
Comment (drewnoakes) [inline src/Aspire.Dashboard/Components/ResourcesGridColumns/SourceColumnDisplay.razor:35]: No longer using `ports`:

```suggestion
else if (Resource.TryGetContainerImage(out var containerImage))
```

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/SourceColumnDisplay.razor
Comment (drewnoakes) [inline src/Aspire.Dashboard/Components/ResourcesGridColumns/SourceColumnDisplay.razor:64]: Can either of `TryGetContainerCommand` or `TryGetContainerArgs` be removed? I don't see any new usages in the code here.

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/UnreadLogErrorsBadge.razor.cs
Comment (drewnoakes) [inline src/Aspire.Dashboard/Components/ResourcesGridColumns/UnreadLogErrorsBadge.razor.cs:18]: Properties like these aren't expected to change over time IIUC. Can these be `init` properties? Ditto elsewhere in the PR.

**Review by drewnoakes — APPROVED** [review]
Approving, but would like to see the `ResourceProperty.DisplayName` used rather than having hard-coded knowledge of different resource properties in the UI.

File: src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs
Comment (JamesNK) [reply src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs:91]: We don't **_need_** to, but it improves the experience. They're a balance between making all data for a resource available, while still having a good default experience.

Known properties server three purposes:
1. By default, display a curated list of information that we think is important. For example, it's useful to know the state, while resource.uid is an internal value.
2. Controlling order of information. Ensures the most useful information is first.
3. Displaying a friendly text in the name column.

This is what it looks like without them:
image

Updating known properties is optional. Data is still available in the UI if not known. Updating UI to show new data is a normal part of app development.

Comment (drewnoakes) [reply src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs:91]: You have to move the loc strings to the server side. All reference data that governs the UI should come from the back end so that we don't have to add explicit support for custom resources in the dashboard.

Comment (drewnoakes) [reply src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs:91]: https://github.com/dotnet/aspire/issues/1644

**Pull request merged**
